hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/FlushTableProcedure.java

```diff
@@ -0,0 +1,178 @@
+        default:
+          throw new UnsupportedOperationException("unhandled state=" + state);
+      }
+    } catch (Exception e) {
+      LOG.warn("Retriable error trying to flush table=" + getTableName() + " state=" + state, e);
+    }
+    return Flow.HAS_MORE_STATE;
+  }
+
+  @Override
+  protected void rollbackState(MasterProcedureEnv env, FlushTableState flushTableState)
+    throws IOException, InterruptedException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  protected FlushTableState getState(int stateId) {
+    return FlushTableState.forNumber(stateId);
+  }
+
+  @Override
+  protected int getStateId(FlushTableState state) {
+    return state.getNumber();
+  }
+
+  @Override
+  protected FlushTableState getInitialState() {
+    return FlushTableState.FLUSH_TABLE_PREPARE;
+  }
+
+  @Override
+  public TableName getTableName() {
+    return tableName;
+  }
+
+  @Override
+  public TableOperationType getTableOperationType() {
+    return TableOperationType.FLUSH;
+  }
+
+  @Override
+  protected void serializeStateData(ProcedureStateSerializer serializer) throws IOException {
+    super.serializeStateData(serializer);
+    FlushTableProcedureStateData.Builder builder = FlushTableProcedureStateData.newBuilder();
+    builder.setTableName(ProtobufUtil.toProtoTableName(tableName));
+    if (columnFamily != null) {
+      builder.setColumnFamily(UnsafeByteOperations.unsafeWrap(columnFamily));
+    }
+    serializer.serialize(builder.build());
+  }
+
+  @Override
+  protected void deserializeStateData(ProcedureStateSerializer serializer) throws IOException {
+    super.deserializeStateData(serializer);
+    FlushTableProcedureStateData data = serializer.deserialize(FlushTableProcedureStateData.class);
+    this.tableName = ProtobufUtil.toTableName(data.getTableName());
+    if (data.hasColumnFamily()) {
+      this.columnFamily = data.getColumnFamily().toByteArray();
+    }
+  }
+
+  private FlushRegionProcedure[] createFlushRegionProcedures(MasterProcedureEnv env) {
+    return env.getAssignmentManager().getTableRegions(getTableName(), true).stream()
+      .filter(r -> RegionReplicaUtil.isDefaultReplica(r))
+      .map(r -> new FlushRegionProcedure(r, columnFamily)).toArray(FlushRegionProcedure[]::new);
+  }
+
+  @Override
+  public void toStringClassDetails(StringBuilder builder) {
+    builder.append(getClass().getName()).append(", id=").append(getProcId()).append(", table=")
+      .append(tableName);
+    if (columnFamily != null) {
+      builder.append(", columnFamily=").append(Bytes.toString(columnFamily));
+    }
+  }
+
+  @Override
+  protected void afterReplay(MasterProcedureEnv env) {
+    if (
+      !env.getMasterConfiguration().getBoolean(
+        MasterFlushTableProcedureManager.FLUSH_PROCEDURE_ENABLED,
+        MasterFlushTableProcedureManager.FLUSH_PROCEDURE_ENABLED_DEFAULT)
+    ) {
+      setFailure("master-flush-table", new IOException("FlushProcedure is DISABLED"));
```

Comment:
The same problem, we have to finish this procedure even if the flag has been changed to disabled after restarting? This procedure does not support rollback...